File: src/main/go/g0001_0100/s0025_reverse_nodes_in_k_group/solution.go
package s0025_reverse_nodes_in_k_group

// #Hard #Top_100_Liked_Questions #Linked_List #Recursion #Data_Structure_II_Day_13_Linked_List
// #Udemy_Linked_List #Top_Interview_150_Linked_List #Big_O_Time_O(n)_Space_O(k)
// #2025_04_27_Time_0_ms_(100.00%)_Space_5.54_MB_(20.53%)

type ListNode struct {
	Val  int
	Next *ListNode
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil || k == 1 {
		return head
	}
	j := 0
	len := head
	// Loop to check the length of the linked list. If the list is less than k, return as it is.
	for j < k {
		if len == nil {
			return head
		}
		len = len.Next
		j++
	}
	// Apply reverse linked list logic.
	c := head
	var n *ListNode
	var prev *ListNode
	i := 0
	// Traverse the while loop for k times to reverse the nodes in k groups.
	for i != k {
		n = c.Next
		c.Next = prev
		prev = c
		c = n
		i++
	}
	// head points to the first node of the k group, which now points to the next k group.
	// Recurse for the remaining linked list.
	head.Next = reverseKGroup(n, k)
	return prev
}

File: src/main/ruby/g0001_0100/s0025_reverse_nodes_in_k_group/solution.rb
# #Hard #Top_100_Liked_Questions #Linked_List #Recursion #Data_Structure_II_Day_13_Linked_List
# #Udemy_Linked_List #Top_Interview_150_Linked_List #Big_O_Time_O(n)_Space_O(k)
# #2023_11_17_Time_62_ms_(91.67%)_Space_211.1_MB_(66.67%)

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @param {Integer} k
# @return {ListNode}
def reverse_k_group(head, k)
  return head if head.nil? || head.next.nil? || k == 1

  j = 0
  len = head

  # Loop for checking the length of the linked list.
  # If the linked list is less than k, then return as it is.
  while j < k
    return head if len.nil?

    len = len.next
    j += 1
  end

  # Reverse linked list logic applied here.
  c = head
  n = nil
  prev = nil
  i = 0

  # Traverse the while loop for K times to reverse the node in K groups.
  while i != k
    n = c.next
    c.next = prev
    prev = c
    c = n
    i += 1
  end

  # head points to the first node of the reversed K group,
  # which is now going to point to the next K group linked list.
  # Recursion for further remaining linked list.
  head.next = reverse_k_group(n, k)
  prev
end
